Bluetooth: Tester: Add Mesh configuration client tests

agata-ponitka opened the pull request.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
 #define MESH_PROXY_IDENTITY		0x13
+#define MESH_CONFIG_BEACON_GET	0x14
+struct mesh_config_val_get{
+	uint16_t address;
```
Net_idx first, address second. Applies to others as well.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
+struct mesh_config_model_pub_get{
+	uint16_t address;
+	uint16_t net_idx;
+	uint16_t el_address;
```
Change el_address to elem_address everywhere.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
+	uint16_t model_ID;
+	uint16_t pub_addr;
+	uint16_t app_idx;
+	bool cred_flag;
```
Use uint8_t here.

michal-narajowski replied:
Ping.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
+struct mesh_config_val_get{
+	uint16_t address;
+	uint16_t net_idx;
+} __packed;
```
Add an empty line between struct and next define.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
 
 #define MESH_RPL_CLEAR			0x12
 #define MESH_PROXY_IDENTITY		0x13
+#define MESH_CONFIG_BEACON_GET	0x14
```
CFG instead of CONFIG

michal-narajowski replied:
in struct name as well

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,6 +741,1163 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+
+	err = bt_mesh_cfg_mod_pub_va_set(cmd->net_idx, cmd->address, cmd->el_address, cmd->model_ID, &pub, &status);
+
+	if (err) {
```
Indentation. Use auto-formatting.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,6 +741,1163 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+
+fail:
+	tester_rsp(BTP_SERVICE_ID_MESH, MESH_CONFIG_MODEL_PUB_VA_SET, CONTROLLER_INDEX, 
+			   err ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS);
```
Indentation. Use auto-formatting.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,6 +741,1163 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+	LOG_DBG("");
+
+	err = bt_mesh_cfg_mod_sub_add(cmd->net_idx, cmd->address, cmd->el_address, 
+								cmd->sub_addr, cmd->model_ID, &status);
```
Indentation. Use auto-formatting.

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,249 @@ struct mesh_lpn_unsubscribe_cmd {
+	uint16_t address;
+	uint16_t net_idx;
+	uint16_t el_address;
+	uint16_t model_ID;
```
mod_id instead of model_ID.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -114,7 +114,12 @@ static void gatt_proxy_status(struct bt_mesh_model *model,
 	state_status_u8(model, ctx, buf, OP_GATT_PROXY_STATUS);
 }
-
+static void transmit_status(struct bt_mesh_model *model,
```
Empty lines needed on both sides.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -713,6 +795,9 @@ const struct bt_mesh_model_op bt_mesh_cfg_cli_op[] = {
 	{ OP_NODE_RESET_STATUS,      0,   node_reset_status},
+	{ OP_NODE_IDENTITY_STATUS,   3,   node_identity_status},
+	{ OP_LPN_TIMEOUT_STATUS,     2,   LPN_timeout_status },
+	{OP_NET_TRANSMIT_STATUS,     1,   transmit_status},
```
Space.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -788,7 +873,7 @@ int bt_mesh_cfg_comp_data_get(uint16_t net_idx, uint16_t addr, uint8_t page,
 	struct bt_mesh_msg_ctx ctx = {
 		.net_idx = net_idx,
-		.app_idx = BT_MESH_KEY_DEV_REMOTE,
+		.app_idx = BT_MESH_KEY_DEV,
```
This change (in all methods) should be a separate commit.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2033,6 +2303,120 @@ int bt_mesh_cfg_hb_pub_get(uint16_t net_idx, uint16_t addr,
+}
+
+int bt_mesh_cfg_node_identity_get(uint16_t net_idx, uint16_t addr, uint16_t key_net_idx, 
+							uint8_t *status, uint8_t *identity)
```
Indentation.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2033,6 +2303,120 @@ int bt_mesh_cfg_hb_pub_get(uint16_t net_idx, uint16_t addr,
+	}
+
+	return cli_wait();
+}
```
Empty line.

michal-narajowski replied:
Empty line missing.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,6 +741,1163 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+	net_buf_simple_init(comp, 0);
+
+	err = bt_mesh_cfg_comp_data_get(cmd->net_idx, cmd->address, cmd->page, &status, comp);
+	if (err) {
```
Indentation.

michal-narajowski reviewed (changes requested):
I have some styling comments and I think this needs to be separated into 3 commits:
1. `BT_MESH_KEY_DEV_REMOTE -> BT_MESH_KEY_DEV` change and other small fixes to existing API.
2. Add missing API.
3. Add support for cfg_cli testing in `tester`.

Other than that, LGTM.

agata-ponitka committed:
Bluetooth: Mesh: Fix Configuration Client API

Hardcoded invalid Company ID was used instead of the parameter passed to
the API.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

agata-ponitka committed:
Bluetooth: Mesh: Add missing Configuration Client API

Add missing API:
- Delete all group addresses in a SIG model's subscription list
- Update a network key
- Update an application key
- Get/Set Node Identity parameters
- Set virtual addtess for a SIG model
- Get/Set Key Refresh Procedures

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

agata-ponitka committed:
Bluetooth: tester: Add Mesh Configuration Client tests support

Add support for automatic testing of Mesh Configuration Client Models.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -923,6 +923,128 @@ int bt_mesh_cfg_hb_pub_set(uint16_t net_idx, uint16_t addr,
+ *  @param net_idx Network index to encrypt with.
+ *  @param addr    Target node address.
+ *  @param unicast_addr LPN unicast address.
+ *	@param polltimeout Poltimeout response parameter.
```
whitespace

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -923,6 +923,128 @@ int bt_mesh_cfg_hb_pub_set(uint16_t net_idx, uint16_t addr,
+ *  @param addr    Target node address.
+ *  @param key_net_idx Network key index the application key belongs to.
+ *  @param status  Status response parameter.
+ *	@param identity Identity response parameter. Must be one of
```
whitespace

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -923,6 +923,128 @@ int bt_mesh_cfg_hb_pub_set(uint16_t net_idx, uint16_t addr,
+ *
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_cfg_LPN_timeout_get(uint16_t net_idx, uint16_t addr, uint16_t unicast_addr, 
```
LPN -> lpn

agata-ponitka committed:
Bluetooth: Mesh: Fix Health Client Model

Assigning a value to the msg_timeout parameter.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

agata-ponitka committed:
Bluetooth: tester: Add Mesh Health Client Model tests support

Add support for automatic testing of Mesh Health Client Models.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

agata-ponitka committed:
Bluetooth: tester: Add provisioner support

Add provisioning procedure for automatic testing.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -78,6 +78,14 @@ int bt_mesh_cfg_comp_data_get(uint16_t net_idx, uint16_t addr, uint8_t page,
+ */
+int bt_mesh_cfg_krp_get(uint16_t net_idx, uint16_t addr, uint16_t key_net_idx, uint8_t *status, uint8_t *phase);
+
+int bt_mesh_cfg_krp_set(uint16_t net_idx, uint16_t addr, uint16_t key_net_idx, uint8_t transition, uint8_t *status, uint8_t *phase);
```
Documentation and code formatting.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -115,6 +115,55 @@ static void gatt_proxy_status(struct bt_mesh_model *model,
+	k_sem_give(&cli->op_sync);
+}
+
+
```
Two empty lines.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -692,6 +742,82 @@ static void hb_pub_status(struct bt_mesh_model *model,
+	k_sem_give(&cli->op_sync);
+}
+
+struct LPN_timeout_param {
```
LPN -> lpn

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -692,6 +742,82 @@ static void hb_pub_status(struct bt_mesh_model *model,
+	uint16_t *polltimeout;
+};
+
+static void LPN_timeout_status(struct bt_mesh_model *model,
```
lpn

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -692,6 +742,82 @@ static void hb_pub_status(struct bt_mesh_model *model,
+			       struct bt_mesh_msg_ctx *ctx,
+			       struct net_buf_simple *buf)
+{
+	struct LPN_timeout_param *param;
```
lpn

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -881,6 +1011,73 @@ int bt_mesh_cfg_beacon_get(uint16_t net_idx, uint16_t addr, uint8_t *status)
+	return cli_wait();
+}
+
+int bt_mesh_cfg_krp_set(uint16_t net_idx, uint16_t addr, uint16_t key_net_idx, uint8_t transition, uint8_t *status, uint8_t *phase)
```
Code formatting.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1161,6 +1399,48 @@ int bt_mesh_cfg_app_key_add(uint16_t net_idx, uint16_t addr, uint16_t key_net_id
+	bt_mesh_model_msg_init(&msg, OP_APP_KEY_UPDATE);
+	key_idx_pack(&msg, key_net_idx, key_app_idx);
+	net_buf_simple_add_mem(&msg, app_key, 16);
+	BT_DBG("0x%04x ", ctx.app_idx);
```
Remaining debug line.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2033,6 +2420,122 @@ int bt_mesh_cfg_hb_pub_get(uint16_t net_idx, uint16_t addr,
+
+	return cli_wait();
+}
+int bt_mesh_cfg_LPN_timeout_get(uint16_t net_idx, uint16_t addr,
```
LPN

michal-narajowski commented on tests/bluetooth/tester/prj.conf:
```diff
@@ -41,4 +41,18 @@ CONFIG_BT_MESH_LPN_POLL_TIMEOUT=100
+CONFIG_BT_MESH_CDB_NODE_COUNT=2
+
 CONFIG_UART_CONSOLE=n
+
```
The following changes are not required.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -83,6 +83,7 @@ static void supported_commands(uint8_t *data, uint16_t len)
 	tester_set_bit(buf->data, MESH_RESET);
 	tester_set_bit(buf->data, MESH_INPUT_NUMBER);
 	tester_set_bit(buf->data, MESH_INPUT_STRING);
+	tester_set_bit(buf->data, MESH_PROVISION_ADV);
```
More commands are missing.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +776,1726 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+	net_buf_simple_init(comp, 0);
+
+	err = bt_mesh_cfg_comp_data_get(cmd->net_idx, cmd->address, cmd->page,
+					&status, comp);
```
We need to rename that argument - it is a page not a status.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +776,1726 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+		goto fail;
+	}
+
+	tester_send(BTP_SERVICE_ID_MESH, MESH_COMP_DATA_GET, CONTROLLER_INDEX,
```
Composition Data Status does not contain a status value.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +776,1726 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+fail:
+	tester_rsp(BTP_SERVICE_ID_MESH, MESH_CFG_RELAY_GET, CONTROLLER_INDEX,
+		   err ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS);
+}
```
Empty line missing.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +776,1726 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+	tester_rsp(BTP_SERVICE_ID_MESH, MESH_CFG_FRIEND_GET, CONTROLLER_INDEX,
+		   err ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS);
+}
+static void config_friend_set(uint8_t *data, uint16_t len)
```
Empty line missing.

michal-narajowski reviewed (changes requested):
Use auto-format tool.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -726,6 +847,10 @@ const struct bt_mesh_model_op bt_mesh_cfg_cli_op[] = {
 	{ OP_HEARTBEAT_SUB_STATUS,   9,   hb_sub_status },
 	{ OP_HEARTBEAT_PUB_STATUS,   10,  hb_pub_status },
 	{ OP_NODE_RESET_STATUS,      0,   node_reset_status },
+	{ OP_NODE_IDENTITY_STATUS,   3,   node_identity_status},
```
Min length does not match with the Spec. Also check others.

michal-narajowski commented on tests/bluetooth/tester/prj.conf:
```diff
@@ -37,8 +37,11 @@ CONFIG_BT_MESH_FRIEND_QUEUE_SIZE=32
 CONFIG_BT_TINYCRYPT_ECC=y
 CONFIG_BT_TESTING=y
 
+
```
Unnecessary line

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -918,6 +918,350 @@ struct mesh_lpn_unsubscribe_cmd {
 #define MESH_RPL_CLEAR			0x12
 #define MESH_PROXY_IDENTITY		0x13
+#define MESH_COMP_DATA_GET		0x14
+struct mesh_comp_data_get {
```
every cmd sould end with _cmd suffix. So in this example `mesh_comp_data_get_cmd`. Look at other structures.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -704,6 +751,80 @@ static void hb_pub_status(struct bt_mesh_model *model,
+	}
+
+	unicast_addr = net_buf_simple_pull_le16(buf);
+	polltimeout = net_buf_simple_pull_le16(buf);
```
Poll Timeout is 3 bytes not 2.

michal-narajowski replied:
Remember to update auto-pts also.

michal-narajowski reviewed (changes requested):
If changes are in `subsys/bluetooth/mesh` then we start a commit message with 
`Bluetooth: Mesh:` and a capital letter. (see other commit messages).

if the change is in `tests/bluetooth/tester` then it's `Bluetooth: tester:`

Missing `Signed-off-by:` in every commit message.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -1470,7 +1470,7 @@ int bt_mesh_cfg_mod_app_get_vnd(uint16_t net_idx, uint16_t addr, uint16_t elem_a
 	}
```
Regarding commit message:

Commit message title for mesh submodule should start with `Bluetooth: Mesh: Fix ...`.

The description should contain more meaningful information, not just repeated title. Missing signed-of-by.

michal-narajowski replied:
Commit description: Hardcoded invalid Company ID was used instead of the parameter passed to the API.

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -926,6 +926,132 @@ int bt_mesh_cfg_hb_pub_set(uint16_t net_idx, uint16_t addr,
 int bt_mesh_cfg_hb_pub_get(uint16_t net_idx, uint16_t addr,
```
Commit title: `Bluetooth: Mesh: Add missing Configuration Client API`

In the description, we need more information, e.g. what API was missing.

Missing signed-of-by.

michal-narajowski commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2021,6 +2297,123 @@ int bt_mesh_cfg_hb_pub_get(uint16_t net_idx, uint16_t addr,
+int bt_mesh_cfg_lpn_timeout_get(uint16_t net_idx, uint16_t addr,
+				uint16_t unicast_addr, int32_t *polltimeout)
+{
+	BT_MESH_MODEL_BUF_DEFINE(msg, OP_LPN_TIMEOUT_GET, 5);
```
Why is the length 5?

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +799,1437 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
-		model_send(data, len);
+	struct mesh_comp_data_get_cmd *cmd = (void *)data;
+	uint8_t page;
+	struct net_buf_simple *comp = NET_BUF_SIMPLE(14);
```
Too small buffer size. It won't store any model data.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -741,53 +799,1437 @@ static void proxy_identity_enable(uint8_t *data, uint16_t len)
+		LOG_ERR("err %d", err);
+		goto fail;
+	}
+
```
Send composition data via `tester_send` to the automation framework. It can be used in other parts of testing.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -2045,6 +2051,214 @@ static void config_lpn_timeout_get(uint8_t *data, uint16_t len)
+{
+	struct mesh_health_fault_test_cmd *cmd = (void *)data;
+	uint8_t faults;
+	size_t fault_count;
```
`faults` should be an array and `fault_count` should contain the number of faults that the application can handle (so the size of the array).

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -2045,6 +2051,214 @@ static void config_lpn_timeout_get(uint8_t *data, uint16_t len)
+	struct mesh_health_fault_clear_cmd *cmd = (void *)data;
+	uint8_t test_id;
+	uint8_t faults;
+	size_t fault_count;
```
`faults` should be an array and `fault_count` should contain the number of faults that the application can handle (so the size of the array).

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -2045,6 +2051,214 @@ static void config_lpn_timeout_get(uint8_t *data, uint16_t len)
+	struct mesh_health_fault_get_cmd *cmd = (void *)data;
+	uint8_t test_id;
+	uint8_t faults;
+	size_t fault_count;
```
faults should be an array and fault_count should contain the number of faults that the application can handle (so the size of the array).

michal-narajowski commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -80,6 +80,32 @@ int bt_mesh_cfg_comp_data_get(uint16_t net_idx, uint16_t addr, uint8_t page,
  */
 int bt_mesh_cfg_beacon_get(uint16_t net_idx, uint16_t addr, uint8_t *status);
 
```
This patch can be squashed with the one adding other missing APIs.

michal-narajowski commented on tests/bluetooth/tester/src/mesh.c:
```diff
@@ -454,7 +459,8 @@ static void config_prov(uint8_t *data, uint16_t len)
 static void provision_node(uint8_t *data, uint16_t len)
 {
-	const struct mesh_provision_node_cmd *cmd = (void *) data;
+	const struct mesh_provision_node_cmd *cmd = (void *)data;
```
These changes regarding provisioner should be merged with previous commit that adds support for provisioner.

michal-narajowski reviewed (changes requested):
Small nitpicks.

agata-ponitka committed:
Bluetooth: tester: Add Key Refresh Procedures tests support

Add support for automatic testing of KRP.

Signed-off-by: Agata Ponitka <agata.ponitka@codecoup.pl>

michal-narajowski commented on tests/bluetooth/tester/src/bttester.h:
```diff
@@ -1247,6 +1248,21 @@ struct mesh_provision_adv_cmd {
 	uint8_t net_key[16];
 } __packed;
 
+#define MESH_CFG_KRP_GET		0x4B
```
Add missing Configuration Client API
